README.md

```diff
@@ -22,10 +22,125 @@ The lack of a standardized, cryptographic verification process for exchange solv
+     * @param amount MUST be the total amount of the asset that the CEX holds on a given chain
+     */
+    struct Asset {
+        string assetName;
+        string chain;
+        uint256 amount;
+    }
+
+    /**
+     * @dev Zero-knowledge proof and its inputs
+     * @param proof MUST be the zero-knowledge proof
+     * @param publicInputs MUST be the proof inputs
+     */
+    struct ZKProof {
+        bytes proof;
+        uint256[] publicInputs;
+    }
+
+    /**
+     * @dev MUST emit when the CEX succesfully submits an address ownership proof
+     * @param addressOwnershipProofs MUST be the list of address ownership proofs
+     */
+    event AddressOwnershipProofSubmitted(
+        AddressOwnershipProof[] addressOwnershipProofs
+    );
+    /**
+     * @dev MUST emit when the CEX succesfully submits a proof of solvency
+     * @param timestamp MUST be the timestamp at which the CEX took the snapshot of its assets and liabilites
+     * @param mstRoot MUST be the Merkle sum tree root of the CEX's liabilities
+     * @param assets MUST be the list of assets owned by the CEX
+     */
+    event SolvencyProofSubmitted(
+        uint256 indexed timestamp,
+        uint256 mstRoot,
+        Asset[] assets
+    );
+
+    /**
+     * @dev Submit an optimistic proof of address ownership for a CEX. The proof is subject to an off-chain verification as it's not feasible to verify the signatures of non-EVM chains in an Ethereum smart contract.
+     * @param _addressOwnershipProofs MUST be the list of address ownership proofs
+     * MUST revert if `_addressOwnershipProofs` array is empty.
+     * MUST revert if any `AddressOwnershipProof` has an empty `cexAddress`, `chain`, `signature`, or `message`.
+     * MUST revert if the `cexAddress` has already been verified.
+     * MUST revert if called by anyone other than the contract owner.
+     * MUST emit `AddressOwnershipProofSubmitted` event upon successful submission.
+     */
+    function submitProofOfAddressOwnership(
+        AddressOwnershipProof[] memory _addressOwnershipProofs
+    ) external;
+
+    /**
+     * @dev Submit proof of solvency of a CEX
+     * @param mstRoot MUST be the Merkle sum tree root of the CEX's liabilities
+     * @param assets MUST be the list of assets owned by the CEX
+     * @param proof MUST be the ZK proof generated from the specified proof of solvency circuit
+     * @param timestamp MUST be the timestamp at which the CEX took the snapshot of its assets and liabilites
+     * In case of successful proof submission `mstRoot` MUST be stored in the contract state against the given `timestamp` because it MAY be used by `verifyProofOfInclusion` function to verify the proof of user inclusion into the Merkle sum tree of the CEX's liabilities. `verifyProofOfInclusion` MUST be able to look up the stored `mstRoot` by a timestamp.
```

Comment:
Should we add this mapping (`timestamp -> mstRoot) to the spec?

Reply:
I think we can, but I'll need to add a reference implementation because a mapping cannot be a part of an interface in Solidity. I'll take a look at how other EIPs are dealing with that. Some have a full reference implementation but I'd prefer to just mention the mapping.